src/safepoint.c

```diff
@@ -124,6 +124,13 @@ int jl_safepoint_start_gc(void)
         jl_safepoint_wait_gc();
         return 0;
     }
+    // Foreign thread adoption disables the GC and waits for it to finish, however, that may
+    // introduce a race between it and this thread checking if the GC is enabled and only
+    // then setting jl_gc_running. To avoid that, check again now that we won that race.
+    if (jl_atomic_load_relaxed(&jl_gc_disable_counter)) {
```

Comment:
Uhm do we have to make this `acquire`

Reply:
I'm not sure. Probably? @vtjnash
If so, I guess the code I copied this from in `jl_gc_collect` should be updated too?

Reply:
I think the issue is that if the check is after the safepoint we have a memory-fence/happens-before-relationship established.

Here we don't have a happens-before relationship with that particular thread since it will not participate in the stop-the world scenario (and this is before the safepoint). 

I tend to go back to https://marabos.nl/atomics/memory-ordering.html#happens-before to refresh my memory.